Support for fuchsia device discovery, workflow, and sdk wrapper

jonahwilliams opened the pull request:
- Adds `FuchsiaSdk` a wrapper/shim which provides access to the set of command line tools necessary to work with a fuchsia device. Internal tool will inject a different implementation that knows where to look in google3.
- Adds `FuchsiaWorkflow` which defines that fuchsia development is supported wherever an `fx` command can be found. This is somewhat hacky...
- Adds `FuchsiaDevices` which adds basic support for device discovery using `netaddr` and `netls` commands. This should work as-is internally when the tool commands are swapped out with a different FuchsiaSdk location.

jonahwilliams committed:
fuchsia devices

jonahwilliams committed:
Add tests and refactor sdk to fx wrapper

jonahwilliams committed:
add missing newlines

jonahwilliams commented:
Woops lol, I should make sure flutter still works...

jonahwilliams committed:
only check if file is null

jonahwilliams committed:
dont look for fuchsia devices

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -27,6 +32,49 @@ class _FuchsiaLogReader extends DeviceLogReader {
+  FuchsiaDevices() : super('Fuchsia devices');
+
+  @override
+  bool get supportsPlatform => true;
```
I don't think we support Windows hosts yet

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -27,6 +32,49 @@ class _FuchsiaLogReader extends DeviceLogReader {
+/// $ ./netls
+/// > device liliac-shore-only-last (fe80::82e4:da4d:fe81:227d/3)
+@visibleForTesting
+void parseFuchsiaDeviceOutput(String text, {@required List<FuchsiaDevice> devices}) {
```
Just make the method return the devices instead of taking the list as a param.

tvolkert replied:
Have `[netls()]` run async and return a `Stream<Uint8List>` (and then maybe make this method an async generator that returns a `Stream<FuchsiaDevice>` instead of a `Future<List<FuchsiaDevice>>`).

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_sdk.dart:
```diff
@@ -0,0 +1,46 @@
+
+/// The fuchsia SDK shell commands.
+///
+/// This workflow assumes development within the fuchsia source tree, providing
```
The wording is a little confusing here (this class doesn't provide the fx command-line tool).  Maybe:

> This workflow assumes development within the Fuchsia source tree, including a working `fx` command-line tool in the user's `PATH`.

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_sdk.dart:
```diff
@@ -0,0 +1,46 @@
+/// The fuchsia SDK.
+FuchsiaSdk get fuchsiaSdk => context[FuchsiaSdk];
+
+/// The fuchsia SDK shell commands.
```
Capitalize Fuchsia?

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_sdk.dart:
```diff
@@ -0,0 +1,46 @@
+class FuchsiaSdk {
+  /// Invokes the `netaddr` command and return the output.
+  ///
+  /// This returns the network address of an attached fuchsia device.
```
What if there are multiple attached devices?

jonahwilliams replied:
I don't really know - I could setup another device and see what the output is, but I was planning on punting on this for now

tvolkert replied:
That's fine - just figured the docs should explicitly state whether it's supported.

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_sdk.dart:
```diff
@@ -0,0 +1,46 @@
+    try {
+      text = runSync(<String>['fx', 'netaddr', '--fuchsia']);
+    } on ArgumentError catch (exception) {
+      throwToolExit('Unable to run "netaddr": ${exception.message}');
```
Is there any useful info outside of `message` such that we should just log `$exception`?

(same comment in `netls()`)

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_sdk.dart:
```diff
@@ -0,0 +1,46 @@
+  /// Invokes the `netaddr` command and return the output.
+  ///
+  /// This returns the network address of an attached fuchsia device.
+  String netaddr() {
```
Run async and return `Stream<Uint8List>`[1]

[1] I know that the declared type of `stdout` is `Stream<List<int>>`, but under the covers, it's guaranteed to be a `Stream<Uint8List>`, and while I understand why it isn't that way[2], it's always nagged at me that the API isn't `Stream<Uint8List>` 😄.  So we may as well cast and expose the modern type here. <span><</span>/soapbox>

[2] afaiui, it's because the `Process` API predates the `Uint8List` type, and we don't want to make the API breaking change at this point.

jonahwilliams replied:
Is there an advantage that `runDetached` has over `runAsync`?  (I'm assuming these are the underlying commands to use). Also attempting to replace runSync (or runAsync) with runDetached/start is returning a process with a `null` stdout and stderr. still looking into why that is

tvolkert replied:
There's a bug to simplify the number of those API methods... I'd stick with runAsync

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_workflow.dart:
```diff
@@ -0,0 +1,34 @@
+
+/// The Fuchsia-specific implementation of a [Workflow].
+///
+/// This workflow assumes development within the fuchsia source tree, providing
```
Same comment as above

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_workflow.dart:
```diff
@@ -0,0 +1,34 @@
+/// the `fx` command-line tool.
+class FuchsiaWorkflow implements Workflow {
+  @override
+  bool get appliesToHostPlatform => true;
```
Not Windows yet

tvolkert commented:
/cc @zanderso

jonahwilliams committed:
Address some review comments

jonahwilliams committed:
fix test

Commit:
git status

Commit:
make fuchsia SDK commands async and add fuchsiaWorkflow to doctor workflows

Commit:
remove hardcoded commands

Commit:
Update fuchsia_device.dart

Commit:
Update fuchsia_device.dart

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -27,6 +33,49 @@ class _FuchsiaLogReader extends DeviceLogReader {
+      continue;
+    // ['device', 'device name', '(id)']
+    final List<String> words = line.split(' ');
+    final String name = words[1];
```
Are we sure the name can't contain spaces?  What about the id?  If the name could but the id can't, then can we assume that everything in between the "device" and the id is the name?

Commit:
Update fuchsia_sdk.dart

Add --nowait flag which removes polling. The flutter tool already handles polling internally

Commit:
Update fuchsia_sdk.dart

testing

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_sdk.dart:
```diff
@@ -0,0 +1,54 @@
+  ///     > fe80::9aaa:fcff:fe60:d3af%eth1
+  Future<String> netaddr() async {
+    try {
+      final RunResult process = await runAsync(<String>['fx', 'netaddr', '--fuchsia', '--nowait']);
```
Just want to note that these commands are likely to be spelled differently at some point.
